### DISCUSSION TITLE
super happy w/ new antidote/prezto setup

### DISCUSSION THREAD

**Discussion by blogabe** [mattmc3/antidote#169] [General]
Hi! I just discovered antidote and am super happy with the decision to switch over. I use Prezto and wanted to clean things up a bit and improve my slow loading experience.

Spent about a week working through the experience and I feel I'm almost there. I have antidote managing prezto directly including my prompt. Same experience as before but speedier and more organized than before.

I do want to test out non prezto plugins and a few different prompts. Spaceship setup seems a pain and maybe not worth the effort. Pure seems nice. PL10k has always been heavy for me.

Thank you so much for creating and maintaining. I really appreciate how this is architected. Working through the hurdles got me exactly what I wanted. Now time to experiment.

**Comment by blogabe** [comment]
If I may ask a quick follow up question... more than anything to ensure I haven't opened a can of worms.

I'm curious why antidote doesn't `source ${ANTIDOTE_HOME:-~}/sorin-ionescu/prezto/init.zsh` when prezto, and only prezto, is being configured in `zsh_plugins.txt`?

I am sourcing this manually before `antidote load`. I get errors otherwise. Thought maybe antidote would do that automatically

**Comment by mattmc3** [comment]
Hey @blogabe! Thank you for the kind words and for using antidote. Glad you've found it to be helpful! I too am a long time Prezto user, and probably wouldn't know what to do without plugins like `sorin-ionescu/prezto path:modules/editor` in my `.zsh_plugins.txt`.

In response to your comments about prompts, might I humbly suggest [using the prompt system from Zephyr?](https://github.com/mattmc3/zephyr) You can get access to all sorts of great prompts with it by simply adding something like this to your Zsh config:

```zsh
# antidote's .zsh_plugins.txt

# Use the 'kind:fpath' annotation to make these prompts available
# to the builtin Zsh prompt system, but no need to actually load them unless you choose them.
# Doing this will not slow down your config.
sindresorhus/pure kind:fpath
romkatv/powerlevel10k kind:fpath
spaceship-prompt/spaceship-prompt kind:fpath

# Now, load an actual prompt plugin to set your prompt
# To set your prompt, add the following zstyle to your .zshrc:
# zstyle ':zephyr:plugin:prompt' theme 'zephyr'
# Valid values are zephyr, starship, spaceship, powerlevel10k, pure, or other prompts you added to your fpath...
mattmc3/zephyr path:plugins/prompt
```

```zsh
# .zshrc

# pick your prompt - zephyr, starship, powerlevel10k, pure
zstyle ':zephyr:plugin:prompt' theme 'zephyr'

# load antidote
source /path/to/antidote.zsh
antidote load
```

> If I may ask a quick follow up question... more than anything to ensure I haven't opened a can of worms. I'm curious why antidote doesn't [source prezto's init.zsh].

You can if you want! Assuming you are using antidote's static file, simply adding this to your .zsh_plugins.txt:

```zsh
# .zsh_plugins.txt
sorin-ionescu/prezto
```

The trouble with doing this is, now you have two different ways to initialize plugins - Prezto's way and antidote's. Prezto requires you to setup all its plugins in zstyle definitions in `.zpreztorc`. Prezto is a full framework and it can be lumbering and slow, but it's simple to setup and use and comes with batteries included. `antidote` lets you mix and match plugins and Zsh scripts as you see fit, but with that comes some complexity if you are mixing and matching subplugins within frameworks (like Oh-My-Zsh and Prezto).

The alternative to letting `Prezto` run your whole config is obviously to let `antidote` do it, and `antidote` lets you then use whatever pieces and parts of `Prezto` or other independent plugins you want. Imagine a config that looks something like this:

```zsh
# .zsh_plugins.txt

# Prezto can be tricky to assemble, so you may need to modify to suit your needs
sorin-ionescu/prezto path:modules/helper/functions kind:autoload
sorin-ionescu/prezto path:modules/helper
sorin-ionescu/prezto path:modules/environment
sorin-ionescu/prezto path:modules/terminal
sorin-ionescu/prezto path:modules/editor
sorin-ionescu/prezto path:modules/history
sorin-ionescu/prezto path:modules/directory
sorin-ionescu/prezto path:modules/spectrum
sorin-ionescu/prezto path:modules/git/functions kind:autoload
sorin-ionescu/prezto path:modules/git
sorin-ionescu/prezto path:modules/history-substring-search
sorin-ionescu/prezto path:modules/history-substring-search
sorin-ionescu/prezto path:modules/completion

# And Prezto has tons more modules...
# sorin-ionescu/prezto path:modules/utility/functions kind:autoload
# sorin-ionescu/prezto path:modules/utility
# sorin-ionescu/prezto path:modules/prompt/functions kind:autoload
# sorin-ionescu/prezto path:modules/prompt
# sorin-ionescu/prezto path:modules/python
# sorin-ionescu/prezto path:modules/python/functions kind:autoload
# sorin-ionescu/prezto path:modules/osx conditional:ismacos
# sorin-ionescu/prezto path:modules/osx/functions kind:autoload conditional:ismacos
# etc...
```

Prezto really does assume that it's run from its own init.zsh, and its subplugins are not independent. That is why I created [Zephyr](https://github.com/mattmc3/zephyr) in the first place - it's a way to get all the goodies from frameworks like Prezto in a loosely coupled way.

Enjoy your Zsh adventure!

**Comment by mattmc3** [comment]
> Spaceship setup seems a pain and maybe not worth the effort.

Forgot to mention this - I think you're right about Spaceship. Zephyr uses [Starship](https://starship.rs) as its prompt (not to be confused with [Spaceship](https://github.com/spaceship-prompt/spaceship-prompt)). While Powerlevel10k is amazing and very fast, it's also massive and can be really complex to configure if you don't want exactly what the wizard produces for you. I like how simple and elegant Starship is, and it's really easy to customize your prompt by [modifying a TOML file](https://github.com/mattmc3/zephyr/blob/main/plugins/prompt/themes/zephyr.toml). It's also [really well documented](https://starship.rs/config/). I have been considering writing a Zsh plugin for Starship that replicates P10k's insta-prompt functionality. That seems to be the main sticking point people have with switching from P10k.

**Comment by blogabe** [comment]
Thanks so much @mattmc3 

I moved all the zstyle definitions from my old `zpreztorc` file into the styles file you use in your example zdotdir and let Antidote handle the module loding. I'm also (still) using the giddie font and source this file directly.

But I have to source `prezto/init.sh` to make it all work otherwise `pmodload` won't run when some of the modules call it.

**Reply by mattmc3** [reply]
> But I have to source prezto/init.sh to make it all work otherwise pmodload won't run when some of the modules call it.

Oh, yeah - `pmodload`. I nearly forgot that I have to have a simple no-op wrapper function for it in my config when using Prezto. If you're using Prezto with antidote, you don't need its `pmodload` function to load your modules - just do it in .zsh_plugins.txt. You can avoid Prezto's init file by simply adding this no-op wrapper function in your .zshrc or in your [autoload functions directory](https://github.com/mattmc3/zfunctions) before loading antidote:

```zsh
pmodload() {}
```

See https://github.com/mattmc3/antidote/discussions/72.

**Reply by blogabe** [reply]
Ah... perfect. Thank you so much. Much cleaner.

**Reply by blogabe** [reply]
I believe then the only additional item I did special because of prezto is changing the location and name of `.zcompdump` to presto's preferred `~/.cache/prezto/zcompdump` when I autoload compinit after adding homebrew to fpath

**Reply by mattmc3** [reply]
Yeah, and there's no way in Prezto to customize that. Hmm... that might merit a PR to Prezto. I've been thinking about doing that for awhile.

**Comment by blogabe** [comment]
Thanks, will take a look at Zephyr

**Comment by blogabe** [comment]
I believe (working to confirm) that this setup is speedy and gets the prezto goodies. I did also set it up so that the zcompfile that's created after homebrew site functions added to fpath uses the same one that prezto uses in `~/.cache/prezto/zcompdump`... additionally utilizing the daily zcompdump caching code. Either way.... so far I'm thrilled with the speed. I have the experience as it used to be just much faster. All zsh config stuff in `~/.config/zsh` and everything cached in `~/..cache/prezto`

**Comment by blogabe** [comment]
Quick follow up here... I'm using Starship, but through antidote not zephyr. I mimicked how you did it meaning setting up a `prompt_starship_setup` for presto to read and added the directory into my fpath. I added the root ZDOTDIR as my module directory and I tried to do it by directly referencing the sub directory, `ZDOTDIR/prompt/functions` The prompt script is simple
```
setopt prompt_subst interactive_comments
export STARSHIP_CONFIG=${ZDOTDIR:-~}/prompt/starship.toml
eval "$(starship init zsh)"
```

However, this doesn't seem to work. I must call `eval "$(starship init zsh)"` directly through `.zshrc` to make this work. That and symlinking the starship prompt directly into $PREZTO_HOME/modules/prompt/functions seem to work fine.

**Reply by mattmc3** [reply]
Here's a bare minimal config I just tried that I have verified works perfectly:

Prompt function file:

```zsh
# $ZDOTDIR/prompt/prompt_mystarship_setup
prompt_opts=(cr percent sp subst)
[[ -r "$1" ]] && export STARSHIP_CONFIG="$1"
eval "$(starship init zsh)"
```

.zshrc

```zsh
# $ZDOTDIR/.zshrc
fpath+=(${ZDOTDIR:-$HOME}/prompt)
autoload -Uz promptinit && promptinit
prompt mystarship ${ZDOTDIR:-$HOME}/prompt/starship.toml
```

**Reply by blogabe** [reply]
Thank you so much. Can confirm this on my end as well.
